### PR TITLE
[ENVIRONMENT] add support for multiple environments

### PR DESCRIPTION
**Pull request by KimForss**
## Problem
The are cases when the deployer is in a different environment from the sap library 

## Solution
Add the support for multiple environment in the naming module, i.e deployer environment and sdu environment

## Tests
<Please provide steps to test the PR>

## Notes
<Additional comments for the PR>

### CODE REVIEW

File: deploy/terraform/terraform-units/modules/sap_namegenerator/variables_local.tf [hunk 1]
```diff
@@ -3,6 +3,16 @@ variable environment {
   description = "Environment type (Prod, Test, Sand, QA)"
 }
 
+variable deployer_environment {
```
Comment (rukawata) [inline deploy/terraform/terraform-units/modules/sap_namegenerator/variables_local.tf:6]: Now that we have environment for deployer and landscape, we need one for `sap_library` and one for `sap_system`?

File: deploy/terraform/terraform-units/modules/sap_namegenerator/resourcegroup.tf [hunk 2]
```diff
@@ -4,15 +4,16 @@ locals {
     upper(format("%s-%s-%s_%s-%s", local.env_verified, local.location_short, local.vnet_verified, var.codename, var.sap_sid))) : (
     upper(format("%s-%s-%s-%s", local.env_verified, local.location_short, local.vnet_verified, var.sap_sid))
   )
-  deployer_name = upper(format("%s-%s-%s", local.env_verified, local.location_short, local.dep_vnet_verified))
-  vnet_name     = upper(format("%s-%s-%s", local.env_verified, local.location_short, local.vnet_verified))
-  library_name  = upper(format("%s-%s", local.env_verified, local.location_short))
+
+  deployer_name   = upper(format("%s-%s-%s", local.deployer_env_verified, local.location_short, local.dep_vnet_verified))
+  landscape_name  = upper(format("%s-%s-%s", local.landscape_env_verified, local.location_short, local.vnet_verified))
```
Comment (rukawata) [inline deploy/terraform/terraform-units/modules/sap_namegenerator/resourcegroup.tf:9]: now that we have separate names for deployer and landscape, can we rename the `vent_verified` to `sap_vnet_verified` (the `dep_vnet_verified` is okay).

File: deploy/terraform/terraform-units/modules/sap_namegenerator/variables_local.tf [hunk 3]
```diff
@@ -274,7 +284,10 @@ locals {
 
   location_short = upper(try(var.region_mapping[var.location], "unkn"))
 
-  env_verified      = upper(substr(var.environment, 0, var.sapautomation_name_limits.environment_variable_length))
+  env_verified            = upper(substr(var.environment, 0, var.sapautomation_name_limits.environment_variable_length))
+  deployer_env_verified   = upper(substr(length(var.deployer_environment) > 0 ? var.deployer_environment : var.environment, 0, var.sapautomation_name_limits.environment_variable_length))
+  landscape_env_verified  = upper(substr(length(var.landscape_environment) > 0 ? var.deployer_environment : var.environment, 0, var.sapautomation_name_limits.environment_variable_length))
```
Comment (rukawata) [inline deploy/terraform/terraform-units/modules/sap_namegenerator/variables_local.tf:289]: I think if we want to support deployer uses deployer environment, landscape uses landscape environment, this overwrites saves some changes in other modules, but makes the logic less clear...Can we do better here?

**Review by rukawata — COMMENTED** [review]
Thanks for the PR. Some suggestions.

File: deploy/terraform/terraform-units/modules/sap_namegenerator/resourcegroup.tf (hunk 2 above)
Comment (KimForss) [reply deploy/terraform/terraform-units/modules/sap_namegenerator/resourcegroup.tf:9]: done

File: deploy/terraform/terraform-units/modules/sap_namegenerator/variables_local.tf (hunk 1 above)
Comment (KimForss) [reply deploy/terraform/terraform-units/modules/sap_namegenerator/variables_local.tf:6]: Not really - unless we want to add it just for clarity? I am ok either way

File: deploy/terraform/terraform-units/modules/sap_namegenerator/variables_local.tf (hunk 3 above)
Comment (KimForss) [reply deploy/terraform/terraform-units/modules/sap_namegenerator/variables_local.tf:289]: Done (maybe :) )

**Review by rukawata — COMMENTED** [review]
Some additional comments. Thanks.

File: deploy/terraform/terraform-units/modules/sap_namegenerator/resourcegroup.tf [hunk 4]
```diff
@@ -1,18 +1,19 @@
...
+  sdu_storageaccount_name            = substr(replace(lower(format("%s%s%sdiag%s", local.env_verified, local.location_short, local.sap_vnet_verified, local.random_id_verified)), "/[^a-z0-9]/", ""), 0, var.azlimits.stgaccnt)
+  library_storageaccount_name        = substr(replace(lower(format("%s%s%ssaplib%s", local.env_verified, local.location_short, local.sap_vnet_verified, local.random_id_verified)), "/[^a-z0-9]/", ""), 0, var.azlimits.stgaccnt)
+  terraformstate_storageaccount_name = substr(replace(lower(format("%s%s%stfstate%s", local.env_verified, local.location_short, local.sap_vnet_verified, local.random_id_verified)), "/[^a-z0-9]/", ""), 0, var.azlimits.stgaccnt)
+  deployer_storageaccount_name       = substr(replace(lower(format("%s%s%sdiag%s", local.deployer_env_verified, local.location_short, local.dep_vnet_verified, local.random_id_verified)), "/[^a-z0-9]/", ""), 0, var.azlimits.stgaccnt)
+  landscape_storageaccount_name      = substr(replace(lower(format("%s%s%sdiag%s", local.landscape_env_verified, local.location_short, local.dep_vnet_verified, local.random_id_verified)), "/[^a-z0-9]/", ""), 0, var.azlimits.stgaccnt)
```
[the remark this answers is not in the document]
Comment (KimForss) [reply deploy/terraform/terraform-units/modules/sap_namegenerator/resourcegroup.tf]: sap_vnet_verified

File: deploy/terraform/terraform-units/modules/sap_namegenerator/variables_local.tf (hunk 1 above)
Comment (KimForss) [reply deploy/terraform/terraform-units/modules/sap_namegenerator/variables_local.tf:11]: added

File: deploy/terraform/terraform-units/modules/sap_namegenerator/resourcegroup.tf [hunk 5]
```diff
@@ -1,18 +1,19 @@
...
+  deployer_storageaccount_name       = substr(replace(lower(format("%s%s%sdiag%s", local.deployer_env_verified, local.location_short, local.dep_vnet_verified, local.random_id_verified)), "/[^a-z0-9]/", ""), 0, var.azlimits.stgaccnt)
+  landscape_storageaccount_name      = substr(replace(lower(format("%s%s%sdiag%s", local.landscape_env_verified, local.location_short, local.sap_vnet_verified, local.random_id_verified)), "/[^a-z0-9]/", ""), 0, var.azlimits.stgaccnt)
+  library_storageaccount_name        = substr(replace(lower(format("%s%ssaplib%s", local.library_env_verified, local.location_short, local.random_id_verified)), "/[^a-z0-9]/", ""), 0, var.azlimits.stgaccnt)
+  sdu_storageaccount_name            = substr(replace(lower(format("%s%s%sdiag%s", local.env_verified, local.location_short, local.sap_vnet_verified, local.random_id_verified)), "/[^a-z0-9]/", ""), 0, var.azlimits.stgaccnt)
+  terraformstate_storageaccount_name = substr(replace(lower(format("%s%s%stfstate%s", local.env_verified, local.location_short, local.sap_vnet_verified, local.random_id_verified)), "/[^a-z0-9]/", ""), 0, var.azlimits.stgaccnt)
```
Comment (YunziZhangAzure) [inline deploy/terraform/terraform-units/modules/sap_namegenerator/resourcegroup.tf:17]: The tfstate storage account should use sap_vnet_verified or library_env_verified? It seems currently the storage account of tfstate is using the environment of saplibrary.

Comment (KimForss) [reply deploy/terraform/terraform-units/modules/sap_namegenerator/resourcegroup.tf:17]: Library - will fix

**Review by YunziZhangAzure — APPROVED** [review]

**Pull request merged**